client/lib/after-layout-flush/index.ts

```diff
@@ -12,18 +14,23 @@
  * https://developer.mozilla.org/en-US/Firefox/Performance_best_practices_for_Firefox_fe_engineers
  *
  * @param {Function} func The function to be invoked after the layout flush
- * @returns {Function} Returns the new delayed function
+ * @returns {Function} The new delayed function
  */
-export default function afterLayoutFlush( func ) {
-	let rafHandle = undefined;
-	let timeoutHandle = undefined;
+export default function afterLayoutFlush< T extends ( ...args: any[] ) => any >(
+	func: T
+): T & Cancelable {
+	// NodeJS and the browser have different types for timeouts (NodeJS.Timeout vs number), so
+	// we can't type this variable, or it will cause typing errors with `clearTimeout` later.
+	let timeoutHandle: any = undefined;
```

Comment:
there's a couple things I could see us doing in a case like this:

Say that we'll use whatever `setTimeout` uses and not think any more about it. I'm not sure how it handles `target: node` vs a browser.

```ts
let timeoutHandle: ReturnType<typeof window.setTimeout>;
```

That can be a little verbose and implementation-specific. A second option, which I think I like better, it to create a name for this value so that the details can be isolated in their own line and so that when people hover over the variable they see something more meaningful than `number`

```ts
type TimerHandle = ReturnType<typeof window.setTimeout>;

let timeout: TimerHandle;
```

![Screen Shot 2019-05-09 at 9 31 44 AM](https://user-images.githubusercontent.com/5431237/57470319-5f5cce80-723d-11e9-9075-7fe766a45a1d.png)

Reply:
Thank you, this does the trick quite well if I go for `ReturnType<typeof setTimeout>`!

Reply:
From my experimentation I'd encourage `ReturnType<typeof window.setTimeout>` because my IDE detected the NodeJS handle instead of the browser one without it

Reply:
Detecting one vs the other isn't an issue, as long as it's consistent. We want it to work in both environments anyway, because of SSR, so I don't have any problem with the IDE detecting the NodeJS version.